semantics/cpp14/language/translation/decl/enum.k

```diff
@@ -0,0 +1,218 @@
+     rule declareEnumerator(E::EnumId, X::CId, prv(_,_,_) #as Pr::PRVal, false, _) => addEnumerator(E, X, Pr)
+
+     // Fixed enumerator.
+     rule declareEnumerator(E::EnumId, X::CId, prv(V::CPPValue, Tr::Trace, _), true, UT::CPPIntegerType) => addEnumerator(E, X, prv(V, Tr, UT))
+     requires inRange(V, UT)
+
+     rule (.K => ILL("TDE1", "Value of enumerator '" +String idToString(X) +String "' does not fit to underlying type.")) ~> declareEnumerator(_, X::CId, prv(V::CPPValue, _, _), true, UT::CPPIntegerType)
+     requires notBool inRange(V, UT)
+
+     syntax KItem ::= addEnumerator(enum: EnumId, id: CId, v: PRVal)
+     rule <k> addEnumerator(E::EnumId, X::CId, V::PRVal) => addEnumeratorToEnv(E, X) ... </k>
+         <enum-id> E </enum-id>
+         <enumerators> ES::Map => ES[X <- V] </enumerators>
+
+     syntax KItem ::= addEnumeratorToEnv(EnumId, CId)
+
+     rule <k> addEnumeratorToEnv(E::EnumId, _) => . ... </k>
+          <enum-id> E </enum-id>
+          <scoped> true </scoped>
+
+     rule <k> addEnumeratorToEnv(enumId(N::Namespace :: Enum(_)) #as EId::EnumId, X::CId) => addUnscopedEnumeratorToNamespace(X, EId, N) ...</k>
+          <curr-tu> Tu::String </curr-tu>
+          <tu-id> Tu </tu-id>
+          <enum-id> EId </enum-id>
+          <scoped> false </scoped>
+
+     syntax KItem ::= addUnscopedEnumeratorToNamespace(CId, EnumId, Namespace)
+
+     rule <k> addUnscopedEnumeratorToNamespace(X::CId, enumId(E::Enum), N::Namespace) => . ... </k>
+          <curr-tu> Tu::String </curr-tu>
+          <tu-id> Tu </tu-id>
+          <ns-id> N </ns-id>
+          <unscoped-enumerators> UE::Map (.Map => X |-> E ) </unscoped-enumerators>
+
+     syntax CPPSimpleType ::= chooseEnumeratorType(CPPSimpleType, Int) [function]
+     rule chooseEnumeratorType(T::CPPSimpleType, Val::Int) => #if inRange(Val, type(T)) #then T #else firstSuitableType(Val, Val, allIntegralTypes) #fi
+
+     // For unscoped enums without fixed type,
+     // compute their minimal and maximal value
+     // as well as their underlying type
+
+     // Possible underlying types according to 7.2:7
+     syntax List ::= "integralTypesGreaterThanInt" [function]
+     rule integralTypesGreaterThanInt
+          => ListItem(int)
+             ListItem(unsigned)
+             ListItem(long)
+             ListItem(unsigned-long)
+             ListItem(long-long)
+             ListItem(unsigned-long-long)
+
+     //
+     syntax List ::= "allIntegralTypes" [function]
+     rule allIntegralTypes
+          => ListItem(signed-char)
+             ListItem(unsigned-char)
+             ListItem(short)
+             ListItem(unsigned-short)
+             integralTypesGreaterThanInt
+
+
+     
+     syntax CPPSimpleType ::= firstSuitableType(Int, Int, List) [function]
+     rule firstSuitableType(Min::Int, Max::Int, ListItem(T::CPPSimpleType) _) => T
+          requires Min >=Int min(type(T)) andBool Max <=Int max(type(T))
+     rule firstSuitableType(_, _, .List) => int
```

Comment:
if we don't expect this case to ever actually occur in well-formed programs, I would prefer to keep the function partial and have it get stuck if somehow we end up at this case by mistake.